cloudinit/config/cc_ansible.py

```diff
@@ -0,0 +1,217 @@
+"""ansible enables running on first boot either ansible-pull"""
+import os
+import re
+import sys
+from copy import deepcopy
+from logging import Logger
+from textwrap import dedent
+from typing import NamedTuple, Optional
+
+from cloudinit.cloud import Cloud
+from cloudinit.config.schema import MetaSchema, get_meta_doc
+from cloudinit.distros import ALL_DISTROS
+from cloudinit.settings import PER_INSTANCE
+from cloudinit.subp import subp, which
+
+meta: MetaSchema = {
+    "id": "cc_ansible",
+    "name": "Ansible",
+    "title": "Configure ansible for instance",
+    "description": dedent(
+        """\
+        This module provides ``ansible`` integration.
+
+        Ansible is often used agentless and in parallel
+        across multiple hosts simultaneously. This
+        doesn't fit the model of cloud-init: a single
+        host configuring itself during boot. Instead,
+        this module installs ansible during boot and
+        then uses ``ansible-pull`` to run the playbook
+        repository at the remote URL.
+        """
+    ),
+    "distros": [ALL_DISTROS],
+    "examples": [
+        dedent(
+            """\
+            #cloud-config
+            ansible:
+              install-method: distro
+              pull:
+                url: "https://github.com/holmanb/vmboot.git"
+                playbook-name: ubuntu.yml
+            """
+        ),
+        dedent(
+            """\
+            #cloud-config
+            ansible:
+              package-name: ansible-core
+              install-method: pip
+              pull:
+                url: "https://github.com/holmanb/vmboot.git"
+                playbook-name: ubuntu.yml
+            """
+        ),
+    ],
+    "frequency": PER_INSTANCE,
+    "activate_by_schema_keys": ["ansible"],
+}
+
+__doc__ = get_meta_doc(meta)
+
+
+class Version(NamedTuple):
+    major: int
+    minor: int
+    patch: int
+
+
+class AnsiblePull:
+    cmd_version: list = []
+    cmd_pull: list = []
+    env = os.environ.copy()
+
+    def get_version(self) -> Optional[Version]:
+        stdout, _ = subp(self.cmd_version, env=self.env)
+        matches = re.search(
+            r"^ansible.*(\d+)\.(\d+).(\d+).*", stdout.splitlines().pop(0)
```

Comment:
I see some older but not ancient Ansible versions having four numeric fields (see `pip index versions ansible`), some where packaged in Debian/Ubuntu (see d/changelog). Also [1] says that ansible-core "does not use semantic versioning" and gives 2.11 and 2.12 as example versions, but I don't see any past version without the PATCH field.

The current code should work even with x.y.z.w versions (ignoring the .w field), but maybe it should be made more general, supporting versions in x.y[.z[.w]] format.

https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html

Reply:
Thanks for the feedback.

I've turned this into a class in util that handles between 2 and 4 digits with comparison operator implementation as a namedtuple subclass (and much better test coverage).

See 5a3c873033b60bd458290bdb746929c28620328c.